Service Worker Fallback Handling

ikreymer opened the pull request:
Adds an injected 'catch all' handler that redirects to the index page if no service worker is installed (#43)

The injected template is in `sw_check.html`.

- If no service worker installed, redirect to index.html with `#redirect=` hash to indicate where to redirect to after installing the SW.
- The index page handles the error message already if service workers are not available.

Note: this is also assumes the raw `text/html;raw=true` support from kiwix/kiwix-lib#403.

The PR addresses:
- #45 - Better redirect from index page, detect prefix by /A/, should fix issue on Android
- #43 - Main SW issue
- #41 - Handles the cold direct links via redirect
- #10 - Handles force refresh
- #23 - 'Dead links' fall back to the kiwix-serve behavior of redirecting to live web or warning, with internal links showing a localized error message (could also be modified).
- #24 - Add additional warning message when loaded over HTTP
- #39 - Add a spinner instead of loading message.

rgaudin commented on src/warc2zim/templates/load.js:
```diff
@@ -6,15 +6,27 @@ async function main() {
-  const parts = window.location.href.split("/");
-  const inx = parts.indexOf("A");
-  const name = parts[inx - 1];
+  var prefix = window.location.href.slice(0, window.location.href.indexOf("/A/"));
```
That won't work on kiwix-serve if it's being used with a `--urlRootLocation` that contains a `/A/`.
Opening a separate issue for this as this might not be trivial to fix.

ikreymer replied:
Added better detection, now using regex to find a `/A/` before a string with `.`, so `http://example.com/A/path/A/example.com` will find `/A/example.com/` not `/A/path/A/example.com`

Of course, can't be entirely full proof.. if someone has `http://example.com/A/example.com/A/example.com/`
then this is ambiguous and not much we can do w/o specifying the prefix explicitly, and I hope that doesn't happen!

rgaudin commented on src/warc2zim/templates/sw_check.html:
```diff
@@ -0,0 +1,22 @@
+<script>
+if (window.top === window && !window._WBWombat) {
+  if (!navigator.serviceWorker || !navigator.serviceWorker.controller) {
+    var inx =  window.location.href.indexOf("/A/");
```
idem

ikreymer commented:
Moving comment here so can see more easily:

Added better detection of prefix, now using regex to find a /A/ before a string with ., so http://example.com/A/path/A/example.com will find /A/example.com/ not /A/path/A/example.com

Of course, can't be entirely full proof.. if someone has http://example.com/A/example.com/A/example.com/
then this is ambiguous and not much we can do w/o specifying the prefix explicitly, and I hope that doesn't happen!

Also added a spinner for #39
